collect_app/src/test/java/org/odk/collect/android/widgets/UrlWidgetTest.java

```diff
@@ -1,46 +1,146 @@
+import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithReadOnly;
+import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.widgetTestActivity;
 
 /**
  * @author James Knight
  */
 
-public class UrlWidgetTest extends QuestionWidgetTest<UrlWidget, StringData> {
-    @NonNull
-    @Override
-    public UrlWidget createWidget() {
-        return new UrlWidget(activity, new QuestionDetails(formEntryPrompt, "formAnalyticsID"));
+@RunWith(RobolectricTestRunner.class)
+public class UrlWidgetTest {
+
+    private CustomTabHelper customTabHelper;
+    private CustomTabsServiceConnection serviceConnection;
+    private OnLongClickListener listener;
+
+    @Before
+    public void setUp() {
+        customTabHelper = mock(CustomTabHelper.class);
+        serviceConnection = mock(CustomTabsServiceConnection.class);
+        listener = mock(OnLongClickListener.class);
+    }
+
+    @Test
+    public void getAnswer_whenPromptAnswerDoesNotHaveAnswer_returnsNull() {
+        assertThat(createWidget(promptWithAnswer(null)).getAnswer(), nullValue());
+    }
+
+    @Test
+    public void getAnswer_whenPromptHasAnswer_returnsAnswer() {
+        UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
+        assertThat(widget.getAnswer().getDisplayText(), equalTo("blah"));
+    }
+
+    @Test
+    public void usingReadOnlyOption_makeAllClickableElementsDisabled() {
+        UrlWidget widget = createWidget(promptWithReadOnly());
+        Button urlButton = widget.findViewById(R.id.url_button);
+        assertThat(urlButton.getVisibility(), equalTo(View.GONE));
+    }
+
+    @Test
+    public void clearAnswer_doesNotClearWidgetAnswer() {
+        UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
+        widget.clearAnswer();
+        assertThat(widget.getAnswer().getDisplayText(), equalTo("blah"));
+    }
+
+    @Test
+    public void whenPromptHasAnswer_displaysAnswer() {
+        UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
+        assertThat(((TextView) widget.findViewById(R.id.url_answer_text)).getText().toString(), equalTo("blah"));
+    }
+
+    @Test
+    public void whenPromptAnswerDoesNotHaveAnswer_displayEmptyString() {
+        UrlWidget widget = createWidget(promptWithAnswer(null));
+        assertThat(((TextView) widget.findViewById(R.id.url_answer_text)).getText().toString(), equalTo(""));
+    }
+
+    @Test
+    public void clickingButtonWhenUrlIsEmpty_doesNotCallOpenUri() {
+        UrlWidget widget = createWidget(promptWithAnswer(null));
+        Button urlButton = widget.findViewById(R.id.url_button);
+        urlButton.performClick();
+
+        verify(customTabHelper, never()).bindCustomTabsService(null, null);
+        verify(customTabHelper, never()).openUri(null, null);
+    }
+
+    @Test
+    public void clickingButton_callsCorrectMethods() {
```

Comment:
I think this could be `_opensUri` instead of `_callsCorrectMethods` so that the intended behavior is clear.